proto/symge/symge-demo.cpp

```diff
@@ -0,0 +1,387 @@
+// bloody CMake
+#ifdef NDEBUG
+#undef NDEBUG
+#endif
+
+#include <cassert>
+#include <cstring>
+#include <iomanip>
+#include <iostream>
+#include <queue>
+#include <random>
+#include <sstream>
+#include <string>
+#include <unordered_set>
+#include <utility>
+
+#include <util/optional.hpp>
+
+#include "symbolic.hpp"
+#include "msparse.hpp"
+
+using namespace sym;
+
+// Identifier name picking helper routines
+
+void join_impl(std::ostream& ss, const std::string& sep) {}
+
+template <typename Head, typename... Args>
+void join_impl(std::ostream& ss, const std::string& sep, Head&& head, Args&&... tail) {
+    if (sizeof...(tail)==0)
+        ss << std::forward<Head>(head);
+    else
+        join_impl(ss << std::forward<Head>(head) << sep, sep, std::forward<Args>(tail)...);
+}
+
+template <typename... Args>
+std::string join(const std::string& sep, Args&&... items) {
+    std::stringstream ss;
+    join_impl(ss, sep, std::forward<Args>(items)...);
+    return ss.str();
+}
+
+class id_maker {
+private:
+    std::unordered_set<std::string> ids;
+
+public:
+    static std::string next_id(std::string s) {
+        unsigned l = s.size();
+        if (l==0) return "a";
+
+        unsigned i = l-1;
+
+        char l0='a', l1='z';
+        char u0='A', u1='Z';
+        char d0='0', d1='9';
+        for (;;) {
+            char& c = s[i];
+            if ((c>=l0 && c<l1) || (c>=u0 && c<u1) || (c>=d0 && c<d1)) {
```

Comment:
Turns out it's a non-standard extension; I'll stick with the `if` as the gain in clarity from a range switch is relatively small.